[FIX] website_slides: fix upload image and pdf

pyduf opened the pull request:
[FIX] website_slides: Fix add image content

When uploading an image content to a course, the image was not uploaded and a
default one was displayed instead. This fixes the problem.

Technical note: The file to upload is converted to base64 when the upload input
is changed and stored in file.data. When the user submit the file, it is that
preprocessed data that is being sent to the server. Before this fix, undefined
was always assigned to file.data because the "dataURL" was splitted in 1
element and the  second element was taken for file.data. The submitted file was
then always empty.

[FIX] website_slides: Fix add pdf content

When uploading a pdf content to a course, the pdf content was not uploaded.
This fixes the problem.

Technical note: The file to upload is converted to base64 when the upload input
is changed and stored in file.data. When the user submit the file, it is that
preprocessed data that is being sent to the server. Following #542cb1dc,
nothing was assigned any more to file.data when the input field was changed so
submitted file was then always empty. Before that commit, the file was read
twice, one for assigning the base64 content of the file to file.data and one to
render the preview. Here we read it only once in base64 to store it in
file.data and then convert it to binary for computing the preview using atob.

[IMP] website_slides: add upload content tests

Test the addition of a new image and a pdf content uploaded from the user
device.

Task-3178726


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/112660).

awa-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,139 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+	},
+}, {
+	content: 'eLearning: check uploaded pdf content',
+	trigger: (backend ? 'iframe ' : '') + '.o_wslides_fs_content.o_wslides_tour_success',
```
Is it not dangerous to check the same class as the "image" part?

Or is the DOM re-initialized at some point which will make them not interfere with each other?

awa-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,139 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+	return expectedContent === actualContent;
+};
+
+const addImageToSection = function (sectionName, pageName, backend) {
```
Also could add a small explanation of what you are testing here specifically (same for the pdf).

awa-odoo reviewed (commented):
Hello @pyduf 

I left some comments and questions.
We could probably squash the 3 commits together as they seem very related (fix preview + associated test).

Thanks & cheers!

pyduf commented:
> Hello @pyduf
> 
> I left some comments and questions. We could probably squash the 3 commits together as they seem very related (fix preview + associated test).
> 
> Thanks & cheers!

Hello @awa-odoo, thanks for the review and sorry for answering so late. Forget to read the general comment, so I will squash the 3 commits with the next changes.

awa-odoo commented on addons/website_slides/static/src/js/slides_upload.js:
```diff
@@ -570,7 +570,7 @@ var SlideUploadDialog = Dialog.extend({
                     preview.src = dataURL;
                 }
-                this.file.data = dataURL.split(',', 1)[1];
+                this.file.data = dataURL.split(',', 2)[1];
```
Ok thanks for the explanation.

As mentioned, can you add a reference to the commit that broke it in the commit message?

Side-note: "splitted" in the commit message should be "split".

awa-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,126 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+};
+
+const checkURLContent = async (url, name, type, expectedContent) => {
+	const blob = await fetch(url).then(res => res.blob());
```
Indent is still strange by the way, seems like a lot more starting space than the previous function.

awa-odoo reviewed (commented):
Hello @pyduf 

Thanks for updating, I added a few extra minor comments.
As mentioned in the first review, I think we could squash the 3 commits.

Cheers!

pyduf commented:
> Hello @pyduf
> 
> Thanks for updating, I added a few extra minor comments. As mentioned in the first review, I think we could squash the 3 commits.
> 
> Cheers!

Hello @awa-odoo, 

Thanks for the review; I have updated the code and squashed the commits.

std-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,126 @@ var addArticleToSection = function (sectionName, pageName, backend) {
 };
 
+const fillInFileInput = async (input, name, type, content) => {
+    const blob = await fetch(`data:${type};base64,${content}`).then(res => res.blob());
```
why do we need to fetch it ? (can't we just base 64 decode it ?)

pyduf replied:
Actually awa-odoo asked me the same question. Although not the straight way to do it, it seems to be the simplest one because following some search on internet, efficient implementation are quite long (convert base64 to blob). Using that tricks is concise and still efficient and it is confined to the tests. I found that solution on stackoverflow. What do you think ?  (note: converted to use only await as suggested below)

std-odoo replied:
yes that's ok so :) it's in a test anyway

std-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,126 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+    return [
+{
+    content: 'eLearning: add content to section',
+    trigger: (backend ? 'iframe ' : '') + 'div.o_wslides_slide_list_category_header:contains("' + sectionName + '") a:contains("Add Content")',
```
can you use "fstring" ? (same below)

pyduf replied:
yes good idea. I have also stored the “backend ? 'iframe ' : ''” in a constant to use in the fstring to avoid the repetition.

std-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,126 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+    content: 'eLearning: check uploaded image presence and perform comparison',
+    trigger: (backend ? 'iframe ' : '') + '.o_wslides_fs_player img',
+    run: async () => {
+        const img = backend ? $('iframe').contents().find('.o_wslides_fs_player img') : $('.o_wslides_fs_player img');
```
can you use document.querySelector instead of jquery ?

pyduf replied:
Changed to use document.querySelector as suggested

std-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,126 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+};
+
+const compareBase64Content = async (url, name, type, expectedContent) => {
+    const blob = await fetch(url).then(res => res.blob());
+    const file = new File([blob], name, { type: type});
+    const actualContent = await utils.getDataURLFromFile(file).then(dataURL => dataURL.split(',', 2)[1]);
```
can you not mix `.then` and `await` ? (you can use `await` everywhere)

std-odoo reviewed (commented):
@pyduf Hi :) 

Some question / suggestion, globally look good :) 

Thanks for your work

pyduf commented:
> @pyduf Hi :)
> 
> Some question / suggestion, globally look good :)
> 
> Thanks for your work

Thanks for the review. I have followed your advice. Only left question is maybe about the fetch I use in the test to create the blob.

std-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,128 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+    content: 'eLearning: check uploaded image presence and perform comparison',
+    trigger: (backend ? 'iframe ' : '') + '.o_wslides_fs_player img',
+    run: async () => {
+        const img = backend ? $('iframe').contents().find('.o_wslides_fs_player img') : $('.o_wslides_fs_player img');
```
still jquery here :(

std-odoo commented on addons/website_slides/static/tests/tours/slides_tour_tools.js:
```diff
@@ -87,6 +95,128 @@ var addArticleToSection = function (sectionName, pageName, backend) {
+        const pdfViewer = baseElement.querySelector('iframe.o_wslides_iframe_viewer').contentDocument
+            .querySelector('#PDFSlideViewer');
+        if (await compareBase64Content(pdfViewer.getAttribute('data-slideurl'), 'Exercise.pdf', 'application/pdf', testPdf)) {
+            const content = backend ? $('iframe').contents().find('.o_wslides_fs_content') : $('.o_wslides_fs_content');
```
same here

std-odoo reviewed (commented):
@pyduf Hi :) 

There's still 2 places with jquery :/

pyduf commented:
> @pyduf Hi :)
> 
> There's still 2 places with jquery :/

Yes, sorry, I was too happy when I had achieved to transform the complex one that I've forgotten the other. I have now converted them all.

std-odoo commented:
@robodoo r+
